File: c#/BigEgg.Algorithm/BigEgg.Algorithm/Sorts/BUMergeSort.cs
using System;
using System.Collections.Generic;

namespace BigEgg.Algorithm.Sorts
{
    public static class BUMergeSort
    {
        /// <summary>
        /// Rearranges the array in ascending order, using the natural order.
        /// </summary>
        /// <param name="a">The array to be sorted</param>
        public static void Sort(IComparable[] a)
        {
            Sort(a, SortOrder.ASC);
        }

        /// <summary>
        /// Rearranges the array in ascending order, using a comparer.
        /// </summary>
        /// <param name="a">The array to be sorted</param>
        /// <param name="c">the comparer that specifying the order</param>
        public static void Sort<Key>(Key[] a, IComparer<Key> c) where Key : class
        {
            Sort(a, c, SortOrder.ASC);
        }

        /// <summary>
        /// Rearranges the array in specific order, using the natural order.
        /// </summary>
        /// <param name="a">The array to be sorted</param>
        /// <param name="order">The sort order.</param>
        public static void Sort(IComparable[] a, SortOrder order)
        {
            if (a == null) { throw new ArgumentNullException("a"); }

            if (order == SortOrder.ASC)
                SortASC(a);
            else
                SortDESC(a);
        }

        /// <summary>
        /// Rearranges the array in specific order, using a comparer.
        /// </summary>
        /// <param name="a">The array to be sorted</param>
        /// <param name="c">the comparer that specifying the order</param>
        /// <param name="order">The sort order.</param>
        public static void Sort<Key>(Key[] a, IComparer<Key> c, SortOrder order) where Key : class
        {
            if (a == null) { throw new ArgumentNullException("a"); }
            if (c == null) { throw new ArgumentNullException("c"); }

            if (order == SortOrder.ASC)
                SortASC(a, c);
            else
                SortDESC(a, c);
        }

        private static void SortASC(IComparable[] a)
        {
            int N = a.Length;
            IComparable[] aux = new IComparable[N];
            for (int sz = 1; sz < N; sz += sz)
                for (int lo = 0; lo < N - sz; lo += sz + sz)
                    MergeASC(a, aux, lo, lo + sz - 1, Math.Min(lo + sz + sz - 1, N - 1));
        }

        private static void MergeASC(IComparable[] a, IComparable[] aux, int lo, int mid, int hi)
        {
            if (!SortHelper.Less(a[mid + 1], a[mid])) return;

            for (int k = lo; k <= hi; k++)
                aux[k] = a[k];

            int i = lo, j = mid + 1;
            for (int k = lo; k <= hi; k++)
            {
                if (i > mid) a[k] = aux[j++];
                else if (j > hi) a[k] = aux[i++];
                else if (SortHelper.Less(aux[j], aux[i])) a[k] = aux[j++];
                else a[k] = aux[i++];
            }
        }

        private static void SortDESC(IComparable[] a)
        {
            int N = a.Length;
            IComparable[] aux = new IComparable[N];
            for (int sz = 1; sz < N; sz += sz)
                for (int lo = 0; lo < N - sz; lo += sz + sz)
                    MergeDESC(a, aux, lo, lo + sz - 1, Math.Min(lo + sz + sz - 1, N - 1));
        }

        private static void MergeDESC(IComparable[] a, IComparable[] aux, int lo, int mid, int hi)
        {
            if (!SortHelper.Greater(a[mid + 1], a[mid])) return;

            for (int k = lo; k <= hi; k++)
                aux[k] = a[k];

            int i = lo, j = mid + 1;
            for (int k = lo; k <= hi; k++)
            {
                if (i > mid) a[k] = aux[j++];
                else if (j > hi) a[k] = aux[i++];
                else if (SortHelper.Greater(aux[j], aux[i])) a[k] = aux[j++];
                else a[k] = aux[i++];
            }
        }


        private static void SortASC<Key>(Key[] a, IComparer<Key> c) where Key : class
        {
            int N = a.Length;
            Key[] aux = new Key[N];
            for (int sz = 1; sz < N; sz += sz)
                for (int lo = 0; lo < N - sz; lo += sz + sz)
                    MergeASC(a, aux, c, lo, lo + sz - 1, Math.Min(lo + sz + sz - 1, N - 1));
        }

        private static void MergeASC<Key>(Key[] a, Key[] aux, IComparer<Key> c, int lo, int mid, int hi) where Key : class
        {
            if (!SortHelper.Less(c, a[mid + 1], a[mid])) return;

            for (int k = lo; k <= hi; k++)
                aux[k] = a[k];

            int i = lo, j = mid + 1;
            for (int k = lo; k <= hi; k++)
            {
                if (i > mid) a[k] = aux[j++];
                else if (j > hi) a[k] = aux[i++];
                else if (SortHelper.Less(c, aux[j], aux[i])) a[k] = aux[j++];
                else a[k] = aux[i++];
            }
        }

        private static void SortDESC<Key>(Key[] a, IComparer<Key> c) where Key : class
        {
            int N = a.Length;
            Key[] aux = new Key[N];
            for (int sz = 1; sz < N; sz += sz)
                for (int lo = 0; lo < N - sz; lo += sz + sz)
                    MergeDESC(a, aux, c, lo, lo + sz - 1, Math.Min(lo + sz + sz - 1, N - 1));
        }

        private static void MergeDESC<Key>(Key[] a, Key[] aux, IComparer<Key> c, int lo, int mid, int hi) where Key : class
        {
            if (!SortHelper.Greater(c, a[mid + 1], a[mid])) return;

            for (int k = lo; k <= hi; k++)
                aux[k] = a[k];

            int i = lo, j = mid + 1;
            for (int k = lo; k <= hi; k++)
            {
                if (i > mid) a[k] = aux[j++];
                else if (j > hi) a[k] = aux[i++];
                else if (SortHelper.Greater(c, aux[j], aux[i])) a[k] = aux[j++];
                else a[k] = aux[i++];
            }
        }
    }
}

File: java/Algorithms/src/Sorts/BUMergeSort.java
package Sorts;

import java.util.Comparator;

import static Sorts.SortHelper.greater;
import static Sorts.SortHelper.less;

/**
 * Created by jianming.xiao on 11/2/14.
 */
public class BUMergeSort {
    /**
     * This class should not be instantiated.
     */
    private BUMergeSort() {
    }

    /**
     * Rearranges the array in ascending order, using the natural order.
     *
     * @param a the array to be sorted
     */
    public static void sort(Comparable[] a) {
        sort(a, SortOrder.ASC);
    }

    /**
     * Rearranges the array in ascending order, using a comparator.
     *
     * @param a the array to be sorted
     * @param c the comparator that specifying the order
     */
    public static void sort(Object[] a, Comparator c) {
        sort(a, c, SortOrder.ASC);
    }

    /**
     * Rearranges the array in specific order, using the natural order.
     *
     * @param a     the array to be sorted
     * @param order the sort order
     */
    public static void sort(Comparable[] a, SortOrder order) {
        if (a == null) throw new IllegalArgumentException();

        if (order == SortOrder.ASC)
            sortASC(a);
        else
            sortDESC(a);
    }

    /**
     * Rearranges the array in specific order, using a comparator.
     *
     * @param a     the array to be sorted
     * @param c     the comparator that specifying the order
     * @param order the sort order
     */
    public static void sort(Object[] a, Comparator c, SortOrder order) {
        if (a == null) throw new IllegalArgumentException();
        if (c == null) throw new IllegalArgumentException();

        if (order == SortOrder.ASC) {
            sortASC(a, c);
        } else {
            sortDESC(a, c);
        }
    }

    protected static void sortASC(Comparable[] a) {
        int N = a.length;
        Comparable[] aux = new Comparable[N];
        for (int sz = 1; sz < N; sz = sz + sz) {
            for (int lo = 0; lo < N - sz; lo += sz + sz) {
                mergeASC(a, aux, lo, lo + sz - 1, Math.min(lo + sz + sz - 1, N - 1));
            }
        }
    }

    protected static void mergeASC(Comparable[] a, Comparable[] aux, int lo, int mid, int hi) {
        if (!less(a[mid + 1], a[mid])) return;

        for (int k = lo; k <= hi; k++) {
            aux[k] = a[k];
        }

        int i = lo, j = mid + 1;
        for (int k = lo; k <= hi; k++) {
            if (i > mid) a[k] = aux[j++];
            else if (j > hi) a[k] = aux[i++];
            else if (less(aux[j], aux[i])) a[k] = aux[j++];
            else a[k] = aux[i++];
        }
    }

    protected static void sortDESC(Comparable[] a) {
        int N = a.length;
        Comparable[] aux = new Comparable[N];
        for (int sz = 1; sz < N; sz = sz + sz) {
            for (int lo = 0; lo < N - sz; lo += sz + sz) {
                mergeDESC(a, aux, lo, lo + sz - 1, Math.min(lo + sz + sz - 1, N - 1));
            }
        }
    }

    protected static void mergeDESC(Comparable[] a, Comparable[] aux, int lo, int mid, int hi) {
        if (!greater(a[mid + 1], a[mid])) return;

        for (int k = lo; k <= hi; k++) {
            aux[k] = a[k];
        }

        int i = lo, j = mid + 1;
        for (int k = lo; k <= hi; k++) {
            if (i > mid) a[k] = aux[j++];
            else if (j > hi) a[k] = aux[i++];
            else if (greater(aux[j], aux[i])) a[k] = aux[j++];
            else a[k] = aux[i++];
        }
    }

    protected static void sortASC(Object[] a, Comparator c) {
        int N = a.length;
        Object[] aux = new Object[N];
        for (int sz = 1; sz < N; sz = sz + sz) {
            for (int lo = 0; lo < N - sz; lo += sz + sz) {
                mergeASC(a, aux, c, lo, lo + sz - 1, Math.min(lo + sz + sz - 1, N - 1));
            }
        }
    }

    protected static void mergeASC(Object[] a, Object[] aux, Comparator c, int lo, int mid, int hi) {
        if (!less(c, a[mid + 1], a[mid])) return;

        for (int k = lo; k <= hi; k++) {
            aux[k] = a[k];
        }

        int i = lo, j = mid + 1;
        for (int k = lo; k <= hi; k++) {
            if (i > mid) a[k] = aux[j++];
            else if (j > hi) a[k] = aux[i++];
            else if (less(c, aux[j], aux[i])) a[k] = aux[j++];
            else a[k] = aux[i++];
        }
    }

    protected static void sortDESC(Object[] a, Comparator c) {
        int N = a.length;
        Object[] aux = new Object[N];
        for (int sz = 1; sz < N; sz = sz + sz) {
            for (int lo = 0; lo < N - sz; lo += sz + sz) {
                mergeDESC(a, aux, c, lo, lo + sz - 1, Math.min(lo + sz + sz - 1, N - 1));
            }
        }
    }

    protected static void mergeDESC(Object[] a, Object[] aux, Comparator c, int lo, int mid, int hi) {
        if (!greater(c, a[mid + 1], a[mid])) return;

        for (int k = lo; k <= hi; k++) {
            aux[k] = a[k];
        }

        int i = lo, j = mid + 1;
        for (int k = lo; k <= hi; k++) {
            if (i > mid) a[k] = aux[j++];
            else if (j > hi) a[k] = aux[i++];
            else if (greater(c, aux[j], aux[i])) a[k] = aux[j++];
            else a[k] = aux[i++];
        }
    }
}
